pimd/pim_cmd_common.c

```diff
@@ -653,3 +663,893 @@ int pim_process_no_rp_plist_cmd(struct vty *vty, const char *rp_str,
+		} else {
+			vty_out(vty,
+				"%-15pPAs %-15pPAs %-16s %-15s %-15s %6d %4d\n",
+				&up->sg.src, &up->sg.grp, rpf_ifname,
+				rpf_addr_str, rib_nexthop_str,
+				rpf->source_nexthop.mrib_route_metric,
+				rpf->source_nexthop.mrib_metric_preference);
+		}
+	}
+}
+
+void pim_show_neighbors_secondary(struct pim_instance *pim, struct vty *vty)
+{
+	struct interface *ifp;
+
+	vty_out(vty,
+		"Interface        Address         Neighbor        Secondary      \n");
+
+	FOR_ALL_INTERFACES (pim->vrf, ifp) {
+		struct pim_interface *pim_ifp;
+		pim_addr ifaddr;
+		struct listnode *neighnode;
+		struct pim_neighbor *neigh;
+
+		pim_ifp = ifp->info;
+
+		if (!pim_ifp)
+			continue;
+
+		if (pim_ifp->pim_sock_fd < 0)
+			continue;
+
+		ifaddr = pim_ifp->primary_address;
+
+		for (ALL_LIST_ELEMENTS_RO(pim_ifp->pim_neighbor_list, neighnode,
+					  neigh)) {
+			struct listnode *prefix_node;
+			struct prefix *p;
+
+			if (!neigh->prefix_list)
+				continue;
+
+			for (ALL_LIST_ELEMENTS_RO(neigh->prefix_list,
+						  prefix_node, p))
+				vty_out(vty,
+					"%-16s %-15pPAs %-15pPAs %-15pFX\n",
+					ifp->name, &ifaddr, &neigh->source_addr,
+					p);
+		}
+	}
+}
+
+void pim_show_state(struct pim_instance *pim, struct vty *vty,
+		    const char *src_or_group, const char *group,
+		    json_object *json)
+{
+	struct channel_oil *c_oil;
+	json_object *json_group = NULL;
+	json_object *json_ifp_in = NULL;
+	json_object *json_ifp_out = NULL;
+	json_object *json_source = NULL;
+	time_t now;
+	int first_oif;
+
+	now = pim_time_monotonic_sec();
+
+	if (!json) {
+		vty_out(vty,
+			"Codes: J -> Pim Join, I -> IGMP Report, S -> Source, * -> Inherited from (*,G), V -> VxLAN, M -> Muted");
+		vty_out(vty,
+			"\nActive Source           Group            RPT  IIF               OIL\n");
+	}
+
+	frr_each (rb_pim_oil, &pim->channel_oil_head, c_oil) {
+		char src_str[PIM_ADDRSTRLEN];
+		char grp_str[PIM_ADDRSTRLEN];
+		char in_ifname[INTERFACE_NAMSIZ + 1];
+		char out_ifname[INTERFACE_NAMSIZ + 1];
+		int oif_vif_index;
+		struct interface *ifp_in;
+		bool isRpt;
+
+		first_oif = 1;
+
+#if PIM_IPV == 4
+		if ((c_oil->up &&
+		     PIM_UPSTREAM_FLAG_TEST_USE_RPT(c_oil->up->flags)) ||
+		    c_oil->oil.mfcc_origin.s_addr == INADDR_ANY)
+			isRpt = true;
+#else
+		if ((c_oil->up &&
+		     PIM_UPSTREAM_FLAG_TEST_USE_RPT(c_oil->up->flags)) ||
+		    pim_addr_is_any(c_oil->oil.mf6cc_origin.sin6_addr))
+			isRpt = true;
+#endif
+		else
+			isRpt = false;
+
+#if PIM_IPV == 4
+		pim_addr mfcc_mcastgrp;
+		pim_addr mfcc_origin;
+
+		memcpy(&mfcc_mcastgrp, &c_oil->oil.mfcc_mcastgrp,
+		       sizeof(struct in_addr));
+		memcpy(&mfcc_origin, &c_oil->oil.mfcc_origin,
+		       sizeof(struct in_addr));
+
+		snprintfrr(grp_str, sizeof(grp_str), "%pPAs", &mfcc_mcastgrp);
+		snprintfrr(src_str, sizeof(src_str), "%pPAs", &mfcc_origin);
+		ifp_in = pim_if_find_by_vif_index(pim, c_oil->oil.mfcc_parent);
+#else
+		snprintfrr(grp_str, sizeof(grp_str), "%pPAs",
+			   &c_oil->oil.mf6cc_mcastgrp);
```

Comment:
Same as above, use `oil_*` functions from `pim_oil.h` and make code the same for IPv4/IPv6.